tensorboard/data/server/cli/dynamic_logdir.rs

```diff
@@ -0,0 +1,91 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+//! Log directory as specified by user arguments.
+
+use log::error;
+use std::collections::HashMap;
+use std::io::{self, Read};
+use std::path::PathBuf;
+
+use crate::disk_logdir::DiskLogdir;
+use crate::gcs;
+use crate::logdir::{EventFileBuf, Logdir};
+use crate::types::Run;
+
+/// A logdir dynamically dispatched over supported implementations.
+pub enum DynLogdir {
+    Disk(DiskLogdir),
+    Gcs(gcs::Logdir),
+}
```

Comment:
@nfelt, re: discussion: you can see from the structure of these types
and implementations how they could be easily macroed into existence—all
except `DynLogdir::new`, of course. It’s basically “the same thing with
more `$` signs”:
<https://gist.github.com/wchargin/f44c370bebb61210dbeb11037f0b2d99>

I think that I have a weak preference for the explicit implementation,
but the macro’s actually not that bad. It does make new cases a one-line
change. Open to suggestions.